### DISCUSSION TITLE
Unable to initialize Firebase from Supabase Edge Function

### DISCUSSION THREAD

**Discussion by rahuldubey11** [supabase/supabase#27856] [Questions]
Hello, I am trying to use Firebase Admin SDK from Supabase Edge Function, but facing issue with correct initialization of the Firebase Admin SDK. Here's my code, followed by the log outputs and my problem description:

### **Code:**

```
import { Database } from './types.ts';
import { initializeApp, cert, ServiceAccount, App, getApps } from "https://esm.sh/firebase-admin@12.2.0/app"
import { getFirestore } from "https://esm.sh/firebase-admin@12.2.0/firestore"
import { getDatabase } from "https://esm.sh/firebase-admin@12.2.0/database"
import { getAuth } from 'https://esm.sh/firebase-admin@12.2.0/auth';

const serviceAccount: ServiceAccount = {
  projectId: Deno.env.get('FIREBASE_PROJECT_ID') as string,
  clientEmail: Deno.env.get('FIREBASE_CLIENT_EMAIL') as string,
  privateKey: Deno.env.get('FIREBASE_PRIVATE_KEY')!.replace(/\\n/g, '\n')
};

let firebaseApp: App;
try {
  firebaseApp = initializeApp({
    credential: cert(serviceAccount as ServiceAccount),
    projectId: Deno.env.get('FIREBASE_PROJECT_ID') as string,
    serviceAccountId: Deno.env.get('FIREBASE_CLIENT_EMAIL') as string,
    databaseURL: Deno.env.get('FIREBASE_DATABASE_URL')
  });
  console.log("Firebase initialized successfully.", {name: firebaseApp.name});

} catch (error) {
  console.error("Failed to initialize Firebase.", error);
}

type TestRecord = Database['public']['Tables'][<my_table>]['Row'];

interface WebhookPayload {
  type: 'INSERT' | 'UPDATE' | 'DELETE' 
  table: string
  record: TestRecord
  schema: 'public'
  old_record: null | TestRecord
}

Deno.serve(async (req) => {
  const payload: WebhookPayload = await req.json();
  const record = payload.record;

  console.log("Initialized Firebase apps count.", {length: getApps().length});

  // ######### FIRESTORE
  let db;
  try {
    db = getFirestore(firebaseApp!);
    console.log("Firestore initialized successfully.", {databaseId: db.databaseId});
  } catch (error) {
    console.error("Failed to initialize Firestore.", error);
  }
  
  // ######### REALTIME DB
  try {
    const realtimeDB = getDatabase(firebaseApp);
    console.log("Realtime DB initialized successfully");
    
    try {
      const rules = await realtimeDB.getRulesJSON();
      console.log("Got rules from Realtime DB.", {rules: rules});

    } catch (error) {
      console.error("Getting data from Realtime DB failed.", error);
    }

  } catch (error) {
    console.error("Realtime DB Error", error);
  }

  // ######### AUTH
  try {
    const user = await getAuth(firebaseApp!).getUser(record.user_id!);
    console.log("Firebase Auth.", {user: user});

  } catch (error) {
    console.error("Firebase Auth Error.", error);
  }

  return new Response('ok');
})
```

### **Log output from above code:**

- Firebase initialized successfully. { name: "[DEFAULT]" }
- Initialized Firebase apps count. { length: 1 }
- _ERROR:_ Failed to initialize Firestore. Error: Failed to initialize Google Cloud Firestore client with the available credentials. Must initialize the SDK with a certificate credential or application default credentials to use Cloud Firestore API.
- Realtime DB initialized successfully
- _ERROR:_ Getting data from Realtime DB failed. Error: secretOrPrivateKey is not valid key material
- _ERROR:_ Firebase Auth Error. Error: secretOrPrivateKey is not valid key material

### **The problem:**
I basically intend to use only Firestore from the edge function, but after failing to initialize it correctly, I have also tried the Real Time Database and Firebase Auth and none of the Firebase SDK are working. As you can see from the logs, no errors are raised when initializeApp() is called, but none of the SDK are getting initialized correctly.

I have spent days searching for similar issues on Github and Q&A forums, tried multiple approaches but still end up with the same errors, can someone help me with this please? Am I doing something wrong, or using Firebase is just not possible with the edge functions?

I have also tried using the service-account.json file directly using
`import serviceAccount from '../service-account.json' with { type: "json" }`
but it also doesn't work as the json file is not being uploaded when the function is deployed, I have tried the approach in official [docs](https://supabase.com/docs/guides/functions/examples/push-notifications?queryGroups=platform&platform=fcm).

Any help would be greatly appreciated.

References: [ref1](https://github.com/GetStream/stream-chat-js/issues/1157), [ref2](https://github.com/firebase/firebase-admin-node/issues/2051), [ref3](https://github.com/supabase/edge-runtime/issues/78#issuecomment-1741302150), [ref4](https://github.com/supabase/edge-runtime/issues/82), [ref5](https://stackoverflow.com/questions/77233075/supabase-sending-push-notification-via-firebase)

**Comment by nyannyacha** [comment]
Hello @rahuldubey093 😁

Instead of using `esm.sh` to load the firebase sdk, please replace it with the npm specifier as shown below.

```typescript
import { initializeApp, cert, ServiceAccount, App, getApps } from "npm:firebase-admin@12.2.0/app"
import { getFirestore } from "npm:firebase-admin@12.2.0/firestore"
import { getDatabase } from "npm:firebase-admin@12.2.0/database"
import { getAuth } from 'npm:firebase-admin@12.2.0/auth';
```

**Reply by rahuldubey11** [reply]
Thank you for the response @nyannyacha. I had already tried using `npm:` and now tried it again. But when deploying the function, I get this error:
<img width="1024" alt="Screenshot 2024-07-09 at 2 06 59 PM" src="https://github.com/supabase/supabase/assets/68503672/fa569c1f-6075-4cfd-9933-cf8ebd77514c">

Do you know how to fix this?
